src/server.rs

```diff
@@ -96,6 +96,65 @@ where
     }
 }
 
+#[cfg(feature = "early-data")]
+impl<IO> TlsStream<IO>
+where
+    IO: AsyncRead + AsyncWrite + Unpin,
+{
+    pub fn poll_read_early_data(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut ReadBuf<'_>,
+    ) -> Poll<io::Result<()>> {
+        use std::io::Read;
+
+        let this = self.get_mut();
+        let mut stream =
+            Stream::new(&mut this.io, &mut this.session).set_eof(!this.state.readable());
+
+        match &this.state {
+            TlsState::Stream | TlsState::WriteShutdown => {
+                {
+                    let mut stream = stream.as_mut_pin();
+
+                    while !stream.eof && stream.session.wants_read() {
+                        match stream.read_io(cx) {
+                            Poll::Ready(Ok(0)) => {
+                                break;
+                            }
+                            Poll::Ready(Ok(_)) => (),
+                            Poll::Pending => {
+                                break;
+                            }
+                            Poll::Ready(Err(err)) => return Poll::Ready(Err(err)),
+                        }
+                    }
+                }
+
+                if let Some(mut early_data) = stream.session.early_data() {
+                    match early_data.read(buf.initialize_unfilled()) {
+                        Ok(n) => {
+                            if n > 0 {
+                                buf.advance(n);
+                                return Poll::Ready(Ok(()));
+                            }
+                        }
+                        Err(err) => return Poll::Ready(Err(err)),
+                    }
+                }
+
+                if stream.session.is_handshaking() {
+                    return Poll::Pending;
```

Comment:
Should we return pending here? or should I handshake? I'm worried about a hang due to missing wake.

Reply:
I don't think there would be a missing wake? since handshake is done once the client sends it's finished, and this code depends on read. I'm not fully sure though.

but now that I think about it, the test code doesn't cover this. since I do `acceptor.accept(&mut sock).await`, the handshake already finishes, so reading the early data "async" effectively did nothing.
should I change some of the server stream logic to use the early data state (kind of like client)? acceptor.accept would no longer wait for the handshake to complete, and writing/reading would need to wait on the handshake to complete.
if I do that, then maybe it would make sense to further the handshake here?

Reply:
If we got `Pending` from `stream.read_io()` it's fine to return `Pending` here, can `is_handshaking()` return `true` after `read_io()` has yielded `Ok(0)`?